### PR TITLE
Add HttpListenerResponse cookie and header tests

### LINKED ISSUE #20159: [porting bug] HttpListenerResponse.SetCookie does not throw in netcoreapp if the Cookie already exists
```cs
[Fact]//[ConditionalFact(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotOneCoreUAP))]
public async Task SetCookie_CookieDoesntExist_ThrowsArgumentException()
{
    HttpListenerResponse response = await GetResponse();
    var cookie1 = new Cookie("name", "value");

    response.SetCookie(cookie1);
    Assert.Throws<ArgumentException>("cookie", () => response.SetCookie(cookie1));

    var cookie2 = new Cookie("name", "value2");
    Assert.Throws<ArgumentException>("cookie", () => response.SetCookie(cookie2));
    Assert.Equal(new Cookie[] { cookie2 }, response.Cookies.Cast<Cookie>());
}
```

Netfx throws an ArgumentException. Netcoreapp does not throw. I have a fix for this and will send a PR when I've written more tests

### LINKED ISSUE #20160: [porting bug] HttpListenerResponse.SetCookies should clone the cookie
The following tests demonstrates this:
```cs
[Fact]
public async Task SetCookie_ValidCookie_ClonesCookie()
{
    HttpListenerResponse response = await GetResponse();
    var cookie = new Cookie("name", "value");
    response.SetCookie(cookie);

    // Cookies are cloned.
    cookie.Value = "value3";
    Assert.Equal("value", response.Cookies[0].Value);
}
```

Netcoreapp fails - `response.Cookies[0].Value == "value3"`
Netfx passes - `response.Cookies[0].Value == "value"`

I have a fix for this which I'll submit with tests

### LINKED ISSUE #20161: [porting bug] HttpListenerResponse.AppendHeaders is broken
```cs


[Fact]//[ConditionalFact(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotOneCoreUAP))]
public async Task AppendHeader_ValidValue_AddsHeaderToCollection()
{
    HttpListenerResponse response = await GetResponse();

    response.AppendHeader("name", "value1");
    Assert.Equal("value1", response.Headers["name"]);

    response.AppendHeader("name", "value2");
    Assert.Equal("value1,value2", response.Headers["name"]);
}

[Fact]//[ConditionalFact(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotOneCoreUAP))]
public async Task AppendHeader_NullOrWhitespaceName_ThrowsArgumentNullException()
{
    HttpListenerResponse response = await GetResponse();
    Assert.Throws<ArgumentNullException>("name", () => response.AppendHeader(null, ""));
}

[Fact]//[ConditionalFact(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotOneCoreUAP))]
public async Task AppendHeader_InvalidNameOrValue_ThrowsArgumentException()
{
    HttpListenerResponse response = await GetResponse();
    Assert.Throws<ArgumentException>("name", () => response.AppendHeader("", ""));
    Assert.Throws<ArgumentException>("name", () => response.AppendHeader("\r \t \n", ""));
    Assert.Throws<ArgumentException>("name", () => response.AppendHeader("(", ""));
    Assert.Throws<ArgumentException>("value", () => response.AppendHeader("name", "value1\rvalue2\r"));
}
```

All these tests pass with netfx. None pass with netcoreapp:
```
   System.Net.Tests.HttpListenerResponseHeadersTests.AppendHeader_NullOrWhitespaceName_ThrowsArgumentNullException [F
  AIL]
        Assert.Equal() Failure
                   (pos 0)
        Expected: name
        Actual:   header
                   (pos 0)
        Stack Trace:
              at System.Net.Tests.HttpListenerResponseHeadersTests.<AppendHeader_NullOrWhitespaceName_ThrowsArgumentNul
  lException>d__4.MoveNext()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
     System.Net.Tests.HttpListenerResponseHeadersTests.AppendHeader_ValidValue_AddsHeaderToCollection [FAIL]
        System.ArgumentException : Specified value does not have a ':' separator.
        Parameter name: header
        Stack Trace:
              at System.Net.WebHeaderCollection.Add(String header)
              at System.Net.HttpListenerResponse.AppendHeader(String name, String value)
              at System.Net.Tests.HttpListenerResponseHeadersTests.<AppendHeader_ValidValue_AddsHeaderToCollection>d__3
  .MoveNext()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
     System.Net.Tests.HttpListenerResponseHeadersTests.AppendHeader_InvalidNameOrValue_ThrowsArgumentException [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentException)
        Actual:   typeof(System.ArgumentNullException): Value cannot be null.
        Parameter name: header
        Stack Trace:
              at System.Net.WebHeaderCollection.Add(String header)
              at System.Net.HttpListenerResponse.AppendHeader(String name, String value)
              at System.Net.Tests.HttpListenerResponseHeadersTests.<>c__DisplayClass5_0.<AppendHeader_InvalidNameOrValu
  e_ThrowsArgumentException>b__0()
```

I have a fix and will submit it with tests

### LINKED ISSUE #20162: [porting bug] HttpListenerResponse crashes with NRE sending any cookies
The following snippet runs to completion with netfx but crashes with NRE with netcoreapp:

```cs
  public static IEnumerable<object[]> Cookies_TestData()
  {
      yield return new object[]
      {
          new CookieCollection() { new Cookie("name", "value") }
      };
  }

  [ConditionalTheory(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotOneCoreUAP))]
  [MemberData(nameof(Cookies_TestData))]
  public async Task Cookies_SetAndSend_ClientReceivesExpectedHeaders(CookieCollection cookies)
  {
      HttpListenerResponse response = await GetResponse();
      response.Cookies = cookies;

      response.Close();
  }
```

Here's the stack trace:
```
System.Net.Tests.HttpListenerResponseCookiesTests.Cookies_SetAndSend_ClientReceivesExpectedHeaders(cookies: [$Vers
  ion=1; name2=value2; $Port=\"300\"]) [FAIL]
        System.NullReferenceException : Object reference not set to an instance of an object.
        Stack Trace:
              at System.Net.CookieExtensions.ToServerString(Cookie cookie)
              at System.Net.HttpListenerResponse.ComputeCookies()
              at System.Net.HttpListenerResponse.ComputeCoreHeaders()
              at System.Net.HttpListenerResponse.ComputeHeaders()
              at System.Net.HttpResponseStream.ComputeLeftToWrite()
              at System.Net.HttpResponseStream.DisposeCore()
              at System.Net.HttpResponseStream.Dispose(Boolean disposing)
              at System.IO.Stream.Close()
              at System.Net.HttpListenerResponse.Dispose()
              at System.Net.HttpListenerResponse.System.IDisposable.Dispose()
              at System.Net.HttpListenerResponse.Close()
              at System.Net.Tests.HttpListenerResponseCookiesTests.<Cookies_SetAndSend_ClientReceivesExpectedHeaders>d_
  _2.MoveNext()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
           --- End of stack trace from previous location where exception was thrown ---
              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
              at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
              at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
```

I have a fix which I'll submit with tests

### PR DESCRIPTION
**Pull request by hughbe**
Contributes to #13618

Fixes #20162
Fixes #20161
Fixes #20160 
Fixes #20159

### CODE REVIEW

File: src/System.Net.HttpListener/src/System/Net/HttpListenerResponse.cs [hunk 1]
```diff
@@ -61,22 +61,15 @@ public void SetCookie(Cookie cookie)
             {
                 throw new ArgumentNullException(nameof(cookie));
             }
-            bool added = false;
-            try
-            {
-                Cookies.Add(cookie);
-                added = true;
-            }
-            catch (CookieException)
```
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/HttpListenerResponse.cs]: Previously this was eating all CookieExceptions.  Is that not being done on desktop?

File: src/System.Net.HttpListener/src/System/Net/Windows/CookieExtensions.cs [hunk 2]
```diff
@@ -15,11 +15,22 @@ public static string ToServerString(this Cookie cookie)
         {
             if (s_toServerStringFunc == null)
             {
-                s_toServerStringFunc = (Func<Cookie, string>)typeof(Cookie).GetMethod("ToServerString").CreateDelegate(typeof(Func<Cookie, string>));
+                s_toServerStringFunc = (Func<Cookie, string>)typeof(Cookie).GetMethod("ToServerString", BindingFlags.NonPublic | BindingFlags.Instance).CreateDelegate(typeof(Func<Cookie, string>));
```
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/Windows/CookieExtensions.cs]: Do we have an issue somewhere about exposing ToServerString?

File: src/System.Net.HttpListener/src/System/Net/Windows/CookieExtensions.cs [hunk 3]
```diff
@@ -15,11 +15,22 @@ public static string ToServerString(this Cookie cookie)
...
             return s_toServerStringFunc(cookie);
         }
 
+        private static Func<Cookie, Cookie> s_cloneFunc;
+
+        public static Cookie Clone(this Cookie cookie)
+        {
+            if (s_cloneFunc == null)
+            {
+                s_cloneFunc = (Func<Cookie, Cookie>)typeof(Cookie).GetMethod("Clone", BindingFlags.NonPublic | BindingFlags.Instance).CreateDelegate(typeof(Func<Cookie, Cookie>));
```
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/Windows/CookieExtensions.cs]: Do we have an issue somewhere tracking exposing Clone (or equivalent functionality)?

File: src/System.Net.HttpListener/src/System/Net/Windows/CookieExtensions.cs [hunk 4]
```diff
@@ -35,12 +46,27 @@ public static bool IsRfc2965Variant(this Cookie cookie)
         {
             if (s_getVariantFunc == null)
             {
-                s_getVariantFunc = (Func<Cookie, CookieVariant>)typeof(Cookie).GetProperty("Variant", BindingFlags.NonPublic).GetGetMethod(true).CreateDelegate(typeof(Func<Cookie, CookieVariant>));
+                s_getVariantFunc = (Func<Cookie, CookieVariant>)typeof(Cookie).GetProperty("Variant", BindingFlags.NonPublic | BindingFlags.Instance).GetGetMethod(true).CreateDelegate(typeof(Func<Cookie, CookieVariant>));
```
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/Windows/CookieExtensions.cs]: Ditto?

File: src/System.Net.HttpListener/src/System/Net/HttpListenerResponse.cs [hunk 5]
```diff
@@ -42,7 +42,7 @@ public void AddHeader(string name, string value)
         public void AppendHeader(string name, string value)
         {
             if (NetEventSource.IsEnabled) NetEventSource.Info(this, $"name={name}, value={value}");
-            Headers.Add(value);
+            Headers.Add(name, value);
```
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/HttpListenerResponse.cs]: I assume this was the cause of test failures in the Windows implementation?

**Comment by stephentoub** [comment]
@dotnet-bot test outerloop netcoreapp Windows_NT Debug
@dotnet-bot test outerloop netcoreapp Ubuntu16.10 Debug

File: src/System.Net.HttpListener/src/System/Net/HttpListenerResponse.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Net.HttpListener/src/System/Net/HttpListenerResponse.cs]: Desktop does not catch anything. In fact I'm fairly sure this `CookieException` catching is dead code. From referencesource:

```
public void Add(Cookie cookie) {
  if (cookie == null) {
      throw new ArgumentNullException("cookie");
  }
  ++m_version;
  int idx = IndexOf(cookie);
  if (idx == -1) {
      m_list.Add(cookie);
  }
  else {
      m_list[idx] = cookie;
  }
}

internal int IndexOf(Cookie cookie) {
    IComparer comp = Cookie.GetComparer();
    int idx = 0;
    foreach (Cookie c in m_list) {
        if (comp.Compare(cookie, c) == 0) {
            return idx;
        }
        ++idx;
    }
    return -1;
}
```

Either way we have to use `CookieCollection.InternalAdd` as it behaves differently for cookies of different variants than `CookieCollection.Add`

File: src/System.Net.HttpListener/src/System/Net/Windows/CookieExtensions.cs (hunk 2 above)
Comment (hughbe) [inline src/System.Net.HttpListener/src/System/Net/Windows/CookieExtensions.cs]: https://github.com/dotnet/corefx/issues/13607 and for the others I'll update the issue

**Comment by stephentoub** [comment]
@dotnet-bot test outerloop netfx Windows_NT Debug

File: src/System.Net.HttpListener/src/System/Net/HttpListenerResponse.cs (hunk 5 above)
Comment (hughbe) [inline src/System.Net.HttpListener/src/System/Net/HttpListenerResponse.cs]: Aye

File: src/System.Net.HttpListener/src/System/Net/Windows/CookieExtensions.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/Windows/CookieExtensions.cs]: Thanks

File: src/System.Net.HttpListener/src/System/Net/HttpListenerResponse.cs (hunk 5 above)
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/HttpListenerResponse.cs]: Got it.

**Pull request merged**